src/ucp/core/ucp_listener.c

```diff
@@ -195,76 +195,134 @@ ucs_status_t ucp_listener_query(ucp_listener_h listener, ucp_listener_attr_t *at
     return UCS_OK;
 }
 
+static void ucp_listener_close_uct_listeners(ucp_listener_h listener)
+{
+    ucp_rsc_index_t i;
+
+    ucs_assert_always(ucp_worker_sockaddr_is_cm_proto(listener->worker));
+
+    for (i = 0; i < listener->cms.num_listeners; ++i) {
+        uct_listener_destroy(listener->cms.listeners[i]);
+    }
+
+    ucs_free(listener->cms.listeners);
+
+    listener->cms.listeners     = NULL;
+    listener->cms.num_listeners = 0;
+}
+
 static void ucp_listener_close_ifaces(ucp_listener_h listener)
 {
+    ucp_worker_h worker;
     int i;
 
-    for (i = 0; i < listener->num_wifaces; i++) {
+    ucs_assert_always(!ucp_worker_sockaddr_is_cm_proto(listener->worker));
+
+    for (i = 0; i < listener->ifaces.num_wifaces; i++) {
+        worker = listener->ifaces.wifaces[i].worker;
+        ucs_assert_always(worker == listener->worker);
         /* remove pending slow-path progress in case it wasn't removed yet */
-        ucs_callbackq_remove_if(&listener->wifaces[i].worker->uct->progress_q,
+        ucs_callbackq_remove_if(&worker->uct->progress_q,
                                 ucp_listener_remove_filter, listener);
-        ucp_worker_iface_cleanup(&listener->wifaces[i]);
+        ucp_worker_iface_cleanup(&listener->ifaces.wifaces[i]);
     }
 
-    ucs_free(listener->wifaces);
+    ucs_free(listener->ifaces.wifaces);
 }
 
-ucs_status_t ucp_listener_create(ucp_worker_h worker,
-                                 const ucp_listener_params_t *params,
-                                 ucp_listener_h *listener_p)
+static ucs_status_t
+ucp_listen_on_cm(ucp_listener_h listener, const ucp_listener_params_t *params)
 {
-    ucp_context_h context   = worker->context;
-    ucp_listener_h listener = NULL;
-    int sockaddr_tls        = 0;
-    char saddr_str[UCS_SOCKADDR_STRING_LEN];
-    ucp_tl_resource_desc_t *resource;
-    uct_iface_params_t iface_params;
-    ucp_worker_iface_t *tmp;
-    ucp_rsc_index_t tl_id;
+    ucp_worker_h worker           = listener->worker;
+    const ucp_rsc_index_t num_cms = ucp_worker_get_cm_num(worker);
```

Comment:
will this function be called if num_cms is zero?

Reply:
no